### PR TITLE
Fix/url shortener

### LINKED ISSUE #880: [FormTribe 🔥][500 Points] URL Shortener
### Welcome to the FormTribe!

This issue will get you **500 points**  for the [FormTribe Hackathon](https://formtribe.com)

**Preparation:**
If this is your first time contributing, please read:

- Our [introductory blog post](https://formbricks.com/blog/join-the-formtribe)
- Overview of [Code Guidelines at Formbricks.](https://formbricks.com/docs/contributing/how-we-code)

**Only when your PR follows _all_ of the best practices it will get merged.**

Want more points quick? Complete [Side Quests 🏰 ](https://formbricks.com/formtribe#side-quests)

----------

### Issue

We allow to add a lot of information via URL parameters which creates long links. 

### Solution

**Solution**: Allow people to use a link shortener

### Tasks

Add a new menu item "Link Shortener" with hero icon "Link" to the dropdown on the right hand. Add it above Developer Docs in the same section (so below the divider):

image

If clicked, open a modal which looks as follows:

image

1. Please use the existing modal component like you can find it for the Create Team functionality.
2. The "Short URL" field you can take from the Share survey modal:

image

Make sure to validate the link before shortening: Only allow links which start with "https://formbricks.com/i/" (SURVEY_BASE_URL from constants) to make sure people don't misuse it for spamming or phishing

Validate on blur with a toast. You can see the UI when you test a URL match in a No Code Action:

image

The toast message should say: Only formbricks links allowed.

Addon: In a different ticket, someone is working on slick link previews. In case the previews are already shipped, make sure that they are still rendered, when the link got shortened.

Use package [nanoId](https://github.com/ai/nanoid) for the new, short URL. Create the short Url server side using server actions. You likely need to add a new database model ShortUrl for storing the long url, short Url pairs. Please do not access prisma directly but build a new service in packages/lib/services. The short Url should look like this:

https://formbricks.com/i/U3ks02kdl2 (length 10, alphanumeric upper+lowercase). Add a new environment variable  SHORT_SURVEY_BASE_URL as well as a constant for this.

Make sure requests to /i/[shortUrl] are redirected properly.

### Thanks!

### PR DESCRIPTION
**Pull request by joyal007**
## What does this PR do?

Fixes #880

Added url shortener feature in formbricks

https://www.loom.com/share/cac9afdd921a43db91674c4e7e09b2f6?sid=d5413a82-a8b6-432e-b234-a9c21e8012af

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change adds a new database migration
- [ ] This change requires a documentation update

## How should this be tested?

- Test A
- Test B

## Checklist

- [ ] Added a screen recording or screenshots to this PR
- [ ] Filled out the "How to test" section in this PR
- [x] Read the [contributing guide](https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md)
- [ ] Self-reviewed my own code
- [ ] Commented on my code in hard-to-understand bits
- [ ] Ran `pnpm build`
- [ ] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [ ] My changes don't cause any responsiveness issues
- [ ] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

**Comment by Dhruwang** [comment]
Thank you @joyal007 for the PR, looks pretty great 💪😊

Here are few issues that I noticed

**Modal not resetting** 
After user closes link shorten modal, its fields should reset

https://github.com/formbricks/formbricks/assets/67850763/24c5f281-b593-4aff-8fc3-caa6a2519983

**Unique URL Pairs**
Ideally it should only have one short URL associated with one long URL, so before creating a new one, we can check whether short url for a given long url already exist 

https://github.com/formbricks/formbricks/assets/67850763/44223847-2d85-4053-9e14-d1a2235e7e7e

We can link icon here too
Screenshot 2023-10-02 at 2 37 18 PM

Rename to [shortUrl]
Screenshot 2023-10-02 at 2 59 49 PM

File: packages/lib/services/shortenedUrl.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { prisma } from "@formbricks/database";
+
+export async function createShortenedUrl(shortUrl: string, longUrl: string) {
```
Comment (Dhruwang) [inline packages/lib/services/shortenedUrl.ts:3]: Add return types to both methods 😊

**Review by Dhruwang — CHANGES_REQUESTED** [review]

File: apps/web/app/(app)/environments/[environmentId]/actions.ts [hunk 2]
```diff
@@ -7,6 +7,19 @@ import { deleteSurvey, getSurvey } from "@formbricks/lib/services/survey";
 import { Team } from "@prisma/client";
 import { Prisma as prismaClient } from "@prisma/client/";
 import { createProduct } from "@formbricks/lib/services/product";
+import { nanoid } from "nanoid";
+import { createShortenedUrl } from "@formbricks/lib/services/shortenedUrl";
+import { SHORT_SURVEY_BASE_URL } from "@formbricks/lib/constants";
+
+export async function createShortUrl(url: string): Promise<string> {
+  const shortUrl = nanoid(10);
```
[the remark this answers is not in the document]
Comment (joyal007) [reply apps/web/app/(app)/environments/[environmentId]/actions.ts]: Thank you bro 
It helped a lot!

**Comment by jobenjada** [comment]
Thanks for your work so far, but as hinted at before, we're merging the first PR to come in on this issue.

Cheers!

**Pull request closed without merging**